source/m2mconnectionhandlerpimpl.cpp

```diff
@@ -373,3 +386,71 @@ void M2MConnectionHandlerPimpl::handle_connection_error(int error)
     //This will come from M2MConnectionSecurity class
     _observer.socket_error(error);
 }
+
+void M2MConnectionHandlerPimpl::init_socket()
+{
+    tr_debug("M2MConnectionHandlerPimpl::init_socket - IN");
+    _error_reported = false;
+    _is_handshaking = false;
+    close_socket();
```

Comment:
For code clarity, close_socket() should not be part of init_socket() function, but should be called explicitly before calling this function. Function name should be reflective of the functions main task , closing socket should be moved out of this function.